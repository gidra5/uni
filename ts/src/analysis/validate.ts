import { Tree } from "../ast";
import { SystemError } from "../error";
import { TokenGroup, ValidatedTokenGroup } from "../parser/tokenGroups";

const validateTokenGroup = (token: TokenGroup): [SystemError[], ValidatedTokenGroup[]] => {
  const errors: SystemError[] = [];

  switch (token.type) {
    case "error": {
      errors.push(token.cause);
      if (!token.token) return [errors, []];

      const [errors2, validatedToken] = validateTokenGroup(token.token);
      errors.push(...errors2);
      return [errors, validatedToken];
    }
    case "group": {
      const [errors2, validatedTokens] = validateTokenGroups(token.tokens);
      errors.push(...errors2);
      return [errors, [{ ...token, tokens: validatedTokens }]];
    }
    default:
      return [errors, [token as ValidatedTokenGroup]];
  }
};

export const validateTokenGroups = (tokens: TokenGroup[]) => {
  return tokens.map(validateTokenGroup).reduce<[SystemError[], ValidatedTokenGroup[]]>(
    ([errors, validated], [errors2, validated2]) => [
      [...errors, ...errors2],
      [...validated, ...validated2],
    ],
    [[], []]
  );
};

const validateChildren = (children: Tree[]) => {
  return children.map(validate).reduce<[SystemError[], Tree[]]>(
    ([errors, validated], [errors2, validated2]) => [
      [...errors, ...errors2],
      [...validated, validated2],
    ],
    [[], []]
  );
};

export const validate = (ast: Tree): [SystemError[], Tree] => {
  const errors: SystemError[] = [];

  switch (ast.type) {
    case "error": {
      errors.push(ast.data.cause);
      if (!ast.children[0]) return [errors, ast];

      const [errors2, validatedTree] = validate(ast.children[0]);
      errors.push(...errors2);
      return [errors, validatedTree];
    }
    default:
      const [errors2, validatedChildren] = validateChildren(ast.children);
      errors.push(...errors2);
      return [errors, { ...ast, children: validatedChildren }];
  }
};
