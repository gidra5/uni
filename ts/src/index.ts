import { program } from "commander";
import readline from "readline";
import { stdin as input, stdout as output } from "process";
import { VM } from "./vm/index.js";
import fs from "fs";
import { parseTokens } from "./parser/tokens.js";
import { parse } from "./parser/index.js";
import { Compiler } from "./compiler/index.js";
import { parseExprString } from "./parser/string.js";
import { evaluate } from "./evaluation/index.js";
import { initialContext } from "./evaluation/utils.js";
import { TaskQueue } from "./evaluation/taskQueue.js";
import { desugar } from "./transformers/desugar.js";
import { identity } from "./utils/index.js";
import { semanticReduction } from "./transformers/semanticReduction.js";

program
  .command("run <file>")
  .description("Run script from a file")
  .action((file) => {
    const taskQueue = new TaskQueue();
    const context = initialContext(taskQueue);
    const code = fs.readFileSync(file, "utf-8");
    const [tokens, tokenErrors] = parseTokens(code);
    const [ast, astErrors] = parse()(tokens);
    const desugared = desugar(ast);
    evaluate(taskQueue, semanticReduction(desugared), context, identity);
    taskQueue.run();
  });

program
  .command("repl [file]")
  .description("Run interactive task queue environment with optional initial script/module")
  .action((file) => {
    const taskQueue = new TaskQueue();
    const context = initialContext(taskQueue);
    if (file) {
      const code = fs.readFileSync(file, "utf-8");
      const [tokens, tokenErrors] = parseTokens(code);
      const [ast, astErrors] = parse()(tokens);
      const desugared = desugar(ast);
      evaluate(taskQueue, semanticReduction(desugared), context, (v) => console.dir(v, { depth: null }));
      taskQueue.run();
    }
    const rl = readline.createInterface({ input, output, prompt: ">> " });
    rl.prompt();

    rl.on("line", (_line) => {
      const line = _line.trim();
      switch (line) {
        case "exit":
          rl.close();
          break;
        default: {
          try {
            const [tokens, tokenErrors] = parseTokens(line);
            const [ast, astErrors] = parse()(tokens);
            // console.dir({ ast, astErrors, tokenErrors }, { depth: null });
            const transformed = desugar(ast);
            // console.dir({ context, transformed }, { depth: null });
            evaluate(taskQueue, transformed, context, (v) => console.dir(v, { depth: null }));
            taskQueue.run();
          } catch (e) {
            console.error(e);
          }
          break;
        }
      }

      rl.prompt();
    }).on("close", () => {
      console.log("Have a great day!");
      process.exit(0);
    });
  });

program
  .command("build <file> <output>")
  .description("Compile a program into an image")
  .action((file, output) => {
    const code = fs.readFileSync(file, "utf-8");
    console.log("File is read");

    // const [tokens, _errors] = parseTokens(code);
    // const [ast, __errors] = parse()(tokens);
    const [ast, _errors] = parseExprString(code);
    console.log("File is parsed");

    const compiled = Compiler.compile(ast, 0x3000);
    console.log("File is compiled");
    const buffer = Buffer.from(compiled.buffer);

    fs.writeFileSync(output, buffer.swap16(), { encoding: "binary" });
    console.log("Compiled output is written to file");
  });

program
  .command("vm <image> [osImage]")
  .description("Run an image of compiled program")
  .action((imageFile, osImageFile) => {
    const osImage = osImageFile ? fs.readFileSync(osImageFile) : undefined;
    const image = fs.readFileSync(imageFile);

    const vm = new VM(osImage);
    console.log("Os image loaded");

    input.setRawMode(true);
    readline.emitKeypressEvents(input);
    input.on("keypress", (char, key) => {
      if (key.ctrl && key.name === "c") {
        process.exit(0);
      }
      vm.emit("input", char);
    });

    vm.loadImage(image);
    console.log("Image loaded");

    vm.on("halt", () => {
      console.log("Halted");

      process.exit(0);
    });

    console.log("Starting VM...");
    vm.run();
  });

program.parse();
