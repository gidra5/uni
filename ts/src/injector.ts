import { FileMap } from "codespan-napi";
import type { Precedence } from "./ast";
import type { Position } from "./position";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  ASTNodeNextId = "ASTNodeNextId",
  ASTNodePrecedenceMap = "ASTNodePrecedenceMap",
  ASTNodePositionMap = "ASTNodePositionMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.ASTNodeNextId]: number;
  [Injectable.ASTNodePrecedenceMap]: Map<string, Precedence>;
  [Injectable.ASTNodePositionMap]: Map<string, Position>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.ASTNodeNextId, 0);
register(Injectable.ASTNodePrecedenceMap, new Map());
register(Injectable.ASTNodePositionMap, new Map());

export { register, inject, Injectable };
