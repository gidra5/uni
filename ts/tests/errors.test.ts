import { beforeEach, expect, it } from "vitest";
import { FileMap } from "codespan-napi";
import { Injectable, register } from "../src/utils/injector";
import { parseTokenGroups } from "../src/parser/tokenGroups";
import { parseScript } from "../src/parser/parser";
import { validate, validateTokenGroups } from "../src/analysis/validate";
import { Tree } from "../src/ast";

beforeEach(() => {
  register(Injectable.NextId, 0);
  register(Injectable.PrecedenceMap, new Map());
  register(Injectable.PositionMap, new Map());
});

function clearIds(ast: Tree) {
  if (ast.children.length > 0) {
    ast.children.forEach(clearIds);
  }
  delete (ast as any).id;
  return ast;
}

const _testCase = (input: string) => {
  const tokens = parseTokenGroups(input);
  // console.dir(tokens, { depth: null });
  const [tokenErrors, validatedTokens] = validateTokenGroups(tokens);
  // console.dir(validatedTokens, { depth: null });
  const ast = parseScript(validatedTokens);
  // console.dir(ast, { depth: null });
  const [errors, validated] = validate(ast);
  // console.dir(validated, { depth: null });

  const fileMap = new FileMap();
  fileMap.addFile("<test>", input);
  const fileId = fileMap.getFileId("<test>");
  for (const error of tokenErrors) error.withFileId(fileId).print(fileMap);
  for (const error of errors) error.withFileId(fileId).print(fileMap);

  expect(tokenErrors.map((e) => e.toObject())).toMatchSnapshot("token errors");
  expect(errors.map((e) => e.toObject())).toMatchSnapshot("errors");
  expect(clearIds(validated)).toMatchSnapshot("ast");
};

it(`single closing parens`, () => _testCase(")"));
it(`single open parens`, () => _testCase("("));
it(`single open parens with number`, () => _testCase("(1"));
it(`single closing brace`, () => _testCase("}"));
it(`single open brace`, () => _testCase("{"));
it(`single closing bracket`, () => _testCase("]"));
it(`single open bracket`, () => _testCase("["));
it(`literal inside unclosed brace inside parens`, () => _testCase("({ 1 )"));
it(`unclosed indexing inside parens`, () => _testCase("(x[1 )"));
it(`literal inside unclosed parens inside braces`, () => _testCase("{ (1 }"));
it(`unclosed indexing inside braces`, () => _testCase("{ x[1 }"));
it(`unclosed parens inside indexing`, () => _testCase("x[(1]"));
it(`literal inside unclosed brace inside indexing`, () => _testCase("x[{ 1 ]"));
it.todo(`num literal application`, () => _testCase("1 2"));
it.todo(`string literal application`, () => _testCase('"1" 2'));
it.todo(`infix operation add no lhs`, () => _testCase("1 +"));
it.todo(`infix operation mult no lhs`, () => _testCase("1 *"));
it.todo(`infix operation mult no rhs`, () => _testCase("* 1"));
it.todo(`infix operation mult no rhs inside parens`, () => _testCase("1 * (* 2)"));
it.todo(`add mult no middle operand`, () => _testCase("1 + * 2"));
it.todo(`add add no middle operand`, () => _testCase("1 + + 2"));
it.todo(`infix operation add long parens no lhs`, () => _testCase("1 + (2 + 3) +"));
it.todo(`infix operation add no lhs`, () => _testCase("1 + 2 + "));
it.todo(`infix operation add unclosed parens`, () => _testCase("1 + (2 + 3"));
it.todo(`single closing parens`, () => _testCase("1 * (5/3) (*4"));
it.todo(`single closing parens`, () => _testCase("send((1+2), 3+,4)"));
it.todo(`single closing parens`, () => _testCase("!"));
it.todo(`single closing parens`, () => _testCase('"\\(")"'));
it.todo(`single closing parens`, () => _testCase("f + !"));
it.only(`single closing parens`, () => _testCase('"uwu\n 1'));
it.todo(`single closing parens`, () => _testCase("1 + (2 + 3))"));
it.todo(`single closing parens`, () => _testCase("1 + (2 + 3)) +"));
it.todo(`single closing parens`, () => _testCase("q + )/"));
it.todo(`single closing parens`, () => _testCase("q + )/"));
it.todo(`single closing parens`, () => _testCase(" - )/q + )/"));

// testCase(
//   `
//     quick_sort := xs -> {
//       if xs == ()
//         return ()

//       (pivot, ...rest) := xs
//       smaller := for x in rest: if x < pivot: x
//       bigger := for x in rest: if x >= pivot: x
//       return (...self smaller, pivot, ...self bigger)
//     }

//     quick_sort (5,3,4,2,1,2)
//   `,
//   [],
//   it.todo
// );
// testCase(
//   `
//     bubble_sort := fn mut list -> {
//       for i in range 0 list.length
//       for j in range 0 (list.length - i - 1)
//       if list[j] > list[j+1]
//         list[j], list[j+1] = list[j+1], list[j]
//       return list
//     }

//     bubble_sort (5,3,4,2,1,2)
//   `,
//   [],
//   it.todo
// );

// testCase(
//   '1 + (2 + 3))',
//   [
//     {
//       message: 'unbalanced parens',
//       cause: [],
//       pos: { start: 7, end: 8 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '1 + (2 + 3)) +',
//   [
//     {
//       message: 'unbalanced parens',
//       cause: [],
//       pos: { start: 7, end: 8 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 9, end: 9 },
//         },
//       ],
//       pos: { start: 8, end: 9 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   ' q + )/',
//   [
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 2, end: 3 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 4, end: 4 },
//         },
//       ],
//       pos: { start: 3, end: 4 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   ' - )/q + )/',
//   [
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 1, end: 2 },
//     },
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 5, end: 6 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 7, end: 7 },
//         },
//       ],
//       pos: { start: 6, end: 7 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   'send(-(2x+7)/A[j, i], 127.0.0.1)',
//   [
//     {
//       message: 'unexpected token inside fn args',
//       cause: [
//         {
//           message: 'unexpected token inside parens',
//           cause: [
//             {
//               message: 'unexpected token: "x"',
//               cause: [],
//               pos: { start: 5, end: 6 },
//             },
//           ],
//           pos: { start: 3, end: 9 },
//         },
//       ],
//       pos: { start: 2, end: 11 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 11, end: 12 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 15, end: 16 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 18, end: 19 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   'send(-(2x+7)/A[j, i], 127.0.0.1 ) + )/',
//   [
//     {
//       message: 'unexpected token inside fn args',
//       cause: [
//         {
//           message: 'unexpected token inside parens',
//           cause: [
//             {
//               message: 'unexpected token: "x"',
//               cause: [],
//               pos: { start: 5, end: 6 },
//             },
//           ],
//           pos: { start: 3, end: 9 },
//         },
//       ],
//       pos: { start: 2, end: 11 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 11, end: 12 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 15, end: 16 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 18, end: 19 },
//     },
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 22, end: 23 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 24, end: 24 },
//         },
//       ],
//       pos: { start: 23, end: 24 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '1 - 1)*1',
//   [
//     {
//       message: 'unexpected closing parenthesis after value',
//       cause: [],
//       pos: { start: 3, end: 4 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '1 - 1)*1+',
//   [
//     {
//       message: 'unexpected closing parenthesis after value',
//       cause: [],
//       pos: { start: 3, end: 4 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 7, end: 7 },
//         },
//       ],
//       pos: { start: 6, end: 7 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '(2x^2-5x+7)-(-i)+ (j++)/0 - )(*f)(2, 7-x, )/q + send(-(2x+7)/A[j, i], 127.0.0.1 ) + )/',
//   [
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: 'unexpected token: "x"',
//           cause: [],
//           pos: { start: 2, end: 3 },
//         },
//       ],
//       pos: { start: 0, end: 11 },
//     },
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: 'unexpected token: "++"',
//           cause: [],
//           pos: { start: 19, end: 20 },
//         },
//       ],
//       pos: { start: 17, end: 21 },
//     },
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 24, end: 25 },
//     },
//     {
//       message: 'missing operator',
//       cause: [],
//       pos: { start: 25, end: 25 },
//     },
//     {
//       message: "symbol can't be used in place of value",
//       cause: [],
//       pos: { start: 26, end: 27 },
//     },
//     {
//       message: 'missing operator',
//       cause: [],
//       pos: { start: 29, end: 29 },
//     },
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: 'unexpected token: ","',
//           cause: [],
//           pos: { start: 31, end: 32 },
//         },
//       ],
//       pos: { start: 29, end: 37 },
//     },
//     {
//       message: 'unexpected token inside fn args',
//       cause: [
//         {
//           message: 'unexpected token inside parens',
//           cause: [
//             {
//               message: 'unexpected token: "x"',
//               cause: [],
//               pos: { start: 45, end: 46 },
//             },
//           ],
//           pos: { start: 43, end: 49 },
//         },
//       ],
//       pos: { start: 42, end: 51 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 51, end: 52 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 55, end: 56 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 58, end: 59 },
//     },
//     {
//       message: 'unexpected closing parenthesis, expected value',
//       cause: [],
//       pos: { start: 62, end: 63 },
//     },
//     {
//       message: 'missing operand',
//       cause: [
//         {
//           message: 'end of tokens',
//           cause: [],
//           pos: { start: 64, end: 64 },
//         },
//       ],
//       pos: { start: 63, end: 64 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '(2^2-5x+7)-(-i)+ (j)/0 - 1)*(1*f)+(27-x, )/q + send(-(2+7)/A,[j, i, 127.0.1 ) + 1)/1',
//   [
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: 'unexpected token: "x"',
//           cause: [],
//           pos: { start: 6, end: 7 },
//         },
//       ],
//       pos: { start: 0, end: 10 },
//     },
//     {
//       message: 'unexpected closing parenthesis after value',
//       cause: [],
//       pos: { start: 23, end: 24 },
//     },
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: 'unexpected token: ","',
//           cause: [],
//           pos: { start: 35, end: 36 },
//         },
//       ],
//       pos: { start: 31, end: 37 },
//     },
//     {
//       message: 'unexpected token inside fn args',
//       cause: [
//         {
//           message: "symbol can't be used in place of value",
//           cause: [],
//           pos: { start: 51, end: 52 },
//         },
//       ],
//       pos: { start: 51, end: 51 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 51, end: 52 },
//     },
//     {
//       message: 'expected comma or closing parens',
//       cause: [],
//       pos: { start: 57, end: 58 },
//     },
//     {
//       message: 'unexpected closing parenthesis after value',
//       cause: [],
//       pos: { start: 61, end: 62 },
//     },
//   ],
//   it.todo
// );

// testCase(
//   '(,) + .. + a',
//   [
//     {
//       message: 'unexpected token inside parens',
//       cause: [
//         {
//           message: "symbol can't be used in place of value",
//           cause: [],
//           pos: { start: 1, end: 2 },
//         },
//         {
//           message: 'unexpected token: ","',
//           cause: [],
//           pos: { start: 1, end: 2 },
//         },
//       ],
//       pos: { start: 0, end: 3 },
//     },
//     {
//       message: "symbol can't be used in place of value",
//       cause: [],
//       pos: { start: 4, end: 5 },
//     },
//   ],
//   it.todo
// );
