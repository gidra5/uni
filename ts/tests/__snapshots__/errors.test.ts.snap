// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`literal inside unclosed brace inside indexing > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "value": "x",
          },
          "type": "name",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "block",
        },
      ],
      "data": {},
      "type": "index",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`literal inside unclosed brace inside indexing > errors 1`] = `[]`;

exports[`literal inside unclosed brace inside indexing > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 6,
        "kind": "primary",
        "message": "must be closed with "}" by that point",
        "start": 6,
      },
      {
        "end": 3,
        "kind": "primary",
        "message": "opening token "{" here",
        "start": 2,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`literal inside unclosed brace inside parens > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "block",
        },
      ],
      "data": {},
      "type": "parens",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`literal inside unclosed brace inside parens > errors 1`] = `[]`;

exports[`literal inside unclosed brace inside parens > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 5,
        "kind": "primary",
        "message": "must be closed with "}" by that point",
        "start": 5,
      },
      {
        "end": 2,
        "kind": "primary",
        "message": "opening token "{" here",
        "start": 1,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`literal inside unclosed parens inside braces > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "parens",
        },
      ],
      "data": {},
      "type": "block",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`literal inside unclosed parens inside braces > errors 1`] = `[]`;

exports[`literal inside unclosed parens inside braces > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 5,
        "kind": "primary",
        "message": "must be closed with ")" by that point",
        "start": 5,
      },
      {
        "end": 3,
        "kind": "primary",
        "message": "opening token "(" here",
        "start": 2,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single closing brace > ast 1`] = `
{
  "children": [],
  "data": {},
  "type": "script",
}
`;

exports[`single closing brace > errors 1`] = `[]`;

exports[`single closing brace > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "unexpected closing token",
        "start": 0,
      },
    ],
    "message": "Unexpected closing token "}"",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single closing bracket > ast 1`] = `
{
  "children": [],
  "data": {},
  "type": "script",
}
`;

exports[`single closing bracket > errors 1`] = `[]`;

exports[`single closing bracket > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "unexpected closing token",
        "start": 0,
      },
    ],
    "message": "Unexpected closing token "]"",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single closing parens > ast 1`] = `
{
  "children": [],
  "data": {},
  "type": "script",
}
`;

exports[`single closing parens > errors 1`] = `[]`;

exports[`single closing parens > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "unexpected closing token",
        "start": 0,
      },
    ],
    "message": "Unexpected closing token ")"",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single open brace > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {},
          "type": "implicit_placeholder",
        },
      ],
      "data": {},
      "type": "block",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`single open brace > errors 1`] = `[]`;

exports[`single open brace > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "must be closed with "}" by that point",
        "start": 1,
      },
      {
        "end": 1,
        "kind": "primary",
        "message": "opening token "{" here",
        "start": 0,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single open bracket > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {},
          "type": "implicit_placeholder",
        },
      ],
      "data": {},
      "type": "square_brackets",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`single open bracket > errors 1`] = `[]`;

exports[`single open bracket > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "must be closed with "]" by that point",
        "start": 1,
      },
      {
        "end": 1,
        "kind": "primary",
        "message": "opening token "[" here",
        "start": 0,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single open parens > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {},
          "type": "implicit_placeholder",
        },
      ],
      "data": {},
      "type": "parens",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`single open parens > errors 1`] = `[]`;

exports[`single open parens > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 1,
        "kind": "primary",
        "message": "must be closed with ")" by that point",
        "start": 1,
      },
      {
        "end": 1,
        "kind": "primary",
        "message": "opening token "(" here",
        "start": 0,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`single open parens with number > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "value": 1,
          },
          "type": "number",
        },
      ],
      "data": {},
      "type": "parens",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`single open parens with number > errors 1`] = `[]`;

exports[`single open parens with number > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 2,
        "kind": "primary",
        "message": "must be closed with ")" by that point",
        "start": 2,
      },
      {
        "end": 1,
        "kind": "primary",
        "message": "opening token "(" here",
        "start": 0,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`unclosed indexing inside braces > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "index",
        },
      ],
      "data": {},
      "type": "block",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`unclosed indexing inside braces > errors 1`] = `[]`;

exports[`unclosed indexing inside braces > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 6,
        "kind": "primary",
        "message": "must be closed with "]" by that point",
        "start": 6,
      },
      {
        "end": 4,
        "kind": "primary",
        "message": "opening token "[" here",
        "start": 3,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`unclosed indexing inside parens > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "index",
        },
      ],
      "data": {},
      "type": "parens",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`unclosed indexing inside parens > errors 1`] = `[]`;

exports[`unclosed indexing inside parens > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 5,
        "kind": "primary",
        "message": "must be closed with "]" by that point",
        "start": 5,
      },
      {
        "end": 3,
        "kind": "primary",
        "message": "opening token "[" here",
        "start": 2,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`unclosed parens inside indexing > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "value": "x",
          },
          "type": "name",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {},
          "type": "parens",
        },
      ],
      "data": {},
      "type": "index",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`unclosed parens inside indexing > errors 1`] = `[]`;

exports[`unclosed parens inside indexing > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 4,
        "kind": "primary",
        "message": "must be closed with ")" by that point",
        "start": 4,
      },
      {
        "end": 3,
        "kind": "primary",
        "message": "opening token "(" here",
        "start": 2,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
]
`;

exports[`unclosed string newline > ast 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "value": "uwu",
      },
      "type": "string",
    },
    {
      "children": [],
      "data": {
        "value": 1,
      },
      "type": "number",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`unclosed string newline > errors 1`] = `[]`;

exports[`unclosed string newline > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 4,
        "kind": "primary",
        "message": "expected closing double quote",
        "start": 4,
      },
    ],
    "message": "Unterminated string literal",
    "notes": [
      "Strings must be enclosed in double quotes (")",
      "Use \\ to escape special characters",
    ],
    "type": "UNTERMINATED_STRING",
  },
]
`;

exports[`unclosed string template > ast 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "value": "uwu",
          },
          "type": "string",
        },
        {
          "children": [],
          "data": {
            "value": 1,
          },
          "type": "number",
        },
      ],
      "data": {},
      "type": "template",
    },
  ],
  "data": {},
  "type": "script",
}
`;

exports[`unclosed string template > errors 1`] = `[]`;

exports[`unclosed string template > token errors 1`] = `
[
  {
    "labels": [
      {
        "end": 9,
        "kind": "primary",
        "message": "must be closed with ")" by that point",
        "start": 9,
      },
      {
        "end": 6,
        "kind": "primary",
        "message": "opening token "\\(" here",
        "start": 4,
      },
    ],
    "message": "Missing closing token",
    "notes": [],
    "type": "MISSING_TOKEN",
  },
  {
    "labels": [
      {
        "end": 9,
        "kind": "primary",
        "message": "expected closing double quote",
        "start": 9,
      },
    ],
    "message": "Unterminated string literal",
    "notes": [
      "Strings must be enclosed in double quotes (")",
      "Use \\ to escape special characters",
    ],
    "type": "UNTERMINATED_STRING",
  },
]
`;
